File: lib/domain/cultivation_plan/dtos/public_plan_save_interaction_rule_reference_row.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Dtos
      class PublicPlanSaveInteractionRuleReferenceRow
        attr_reader :reference_interaction_rule_id,
                    :rule_type,
                    :source_group,
                    :target_group,
                    :impact_ratio,
                    :is_directional,
                    :region,
                    :description

        # @param reference_interaction_rule_id [Integer, #to_i]
        # @param rule_type [String, #to_s]
        # @param source_group [String, #to_s]
        # @param target_group [String, #to_s]
        # @param impact_ratio [Numeric]
        # @param is_directional [Boolean]
        # @param region [String, nil]
        # @param description [String, nil]
        def initialize(
          reference_interaction_rule_id:,
          rule_type:,
          source_group:,
          target_group:,
          impact_ratio:,
          is_directional:,
          region: nil,
          description: nil
        )
          @reference_interaction_rule_id = reference_interaction_rule_id.to_i
          @rule_type = rule_type.to_s
          @source_group = source_group.to_s
          @target_group = target_group.to_s
          @impact_ratio = impact_ratio
          @is_directional = !!is_directional
          @region = region.nil? ? nil : region.to_s
          @description = description
          freeze
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/dtos/public_plan_save_interaction_rule_reference_row.rs
//! Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSaveInteractionRuleReferenceRow`

/// Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSaveInteractionRuleReferenceRow`
#[derive(Debug, Clone)]
pub struct PublicPlanSaveInteractionRuleReferenceRow {
    pub reference_interaction_rule_id: i64,
    pub rule_type: String,
    pub source_group: String,
    pub target_group: String,
    pub impact_ratio: f64,
    pub is_directional: bool,
    pub region: Option<String>,
    pub description: Option<String>,
}

impl PublicPlanSaveInteractionRuleReferenceRow {
    pub fn new(
        reference_interaction_rule_id: i64,
        rule_type: impl Into<String>,
        source_group: impl Into<String>,
        target_group: impl Into<String>,
        impact_ratio: f64,
        is_directional: bool,
        region: Option<String>,
        description: Option<String>,
    ) -> Self {
        Self {
            reference_interaction_rule_id,
            rule_type: rule_type.into(),
            source_group: source_group.into(),
            target_group: target_group.into(),
            impact_ratio,
            is_directional,
            region,
            description,
        }
    }
}
